feat: add custom installation loader to the install script

varshith257 opened the pull request:
## Description
Added a combined progress bar and spinner to visually represent the download progress, displaying both percentage completion and a graphical progress bar.


## Related Issue(s)

This PR addresses issue #970
/claim #970

## Screenshots
```
(zulip-py3-venv) vamshi@LAPTOP-L0PR20E0:/mnt/c/Users/91970/Desktop/algora/daytona$ sudo hack/install.sh
Installing Daytona...

Default installation directory: /usr/local/bin
You can override this by setting the DAYTONA_PATH environment variable (ie. `| DAYTONA_PATH=/home/user/bin bash`)

Downloading Daytona binary from https://download.daytona.io/daytona/latest/daytona-linux-amd64
[⠏] Downloading... [####################################----] 91%
```
## Notes
To test it use:

```
curl -sf -L https://raw.githubusercontent.com/varshith257/daytona/add-custom-install-loader/hack/install.sh | sudo bash
```

varshith257 commented:
```
Default installation directory: /usr/local/bin
You can override this by setting the DAYTONA_PATH environment variable (ie. `| DAYTONA_PATH=/home/user/bin bash`)

Downloading Daytona binary from https://download.daytona.io/daytona/latest/daytona-linux-amd64
[⠸] Downloading... [########################################-] 100%
Installing server to /usr/local/bin

Daytona has been successfully installed to /usr/local/bin!
```

Tpuljak commented on hack/install.sh:
```diff
@@ -134,6 +134,45 @@ temp_file="daytona-$RANDOM"
+  local temp_output=$(mktemp)
+
+  # Start downloading the file with `curl` and pipe the output to `dd` to track progress
+  curl -L -s "$url" -o "$temp_output" &
```
![Screenshot 2024-08-28 at 15 13 59](https://github.com/user-attachments/assets/30834dcb-e47e-4575-a756-a6b0e141b9c2)

@varshith257 any reason why you decided to skip `-S` and `-f`? There are descriptions for those flags in the comment below.

P.S. That comment should also be moved up here.

varshith257 commented:
@Tpuljak Thanks for pointing that out! Added back it

The primary reason for not including the -S and -f flags is that the progress bar logic I have done with curl default --progress bar flag and including -S or -f had interfered with capturing and displaying the download progress correctly. So instead I implemented custom loader

Tpuljak reviewed (changes requested):
Please also add

`curl -sf -L https://raw.githubusercontent.com/varshith257/daytona/add-custom-install-loader/hack/install.sh | sudo bash` in the description of the PR so other maintainers have an easier time testing this.

varshith257 commented:
Interesting! I think the stat command in macOS has a different syntax compared to Linux which is why we see errors like the illegal option

varshith257 commented:
Yes! Just added have a look at it

Tpuljak commented on hack/install.sh:
![Screenshot 2024-08-29 at 08 55 08](https://github.com/user-attachments/assets/22a4ae5e-c774-48df-8ece-d2cf2acc34b3)

If the previous progress bar was longer than the next, there are some leftovers in the terminal (e.g. you can see a double % printed above). Please address that

varshith257 replied:
Fixed

Tpuljak reviewed (approved):
Nice work!

Tpuljak commented:
Just a note for you for future commits. Please include the subject (`feat:`/`fix:` in the commit message as well. I can see it in the title of the PR but not in the commit.

Not important for this PR, but keep that in mind please.

idagelic reviewed (commented):
Hi, just tried the installer and it was successful but got "stuck" at 99% - can you take a look?

![image](https://github.com/user-attachments/assets/45fbb424-deda-469b-aeaf-8e7332b8fee5)

varshith257 commented:
Okay! I will test it again and update here

varshith257 commented:
@idagelic Just fixed it! The problem occurred due to synchronization of the percentage bar with the spinner. I have ran multiple times and getting consistent results:
![image](https://github.com/user-attachments/assets/9ad49447-b1d5-47df-8ff1-51cb4727762e)

Tpuljak reviewed (changes requested):
![Screenshot 2024-08-30 at 16 04 57](https://github.com/user-attachments/assets/c1002251-6ce7-4f50-9501-97419cc5ec20)

@varshith257 now I'm getting 99% as well 😅 

Maybe if you manually write the full progress bar at the end if it's successful?

idagelic reviewed (approved):
Works well now, thanks!

Tpuljak reviewed (approved):
Nice work!
